mailscanner/user_manager.php

```diff
@@ -55,6 +55,482 @@ function getHtmlMessage($value, $type)
+    $sql = "SELECT filter FROM user_filters WHERE username = '" . $_SESSION['myusername'] . "'";
+    $result = dbquery($sql);
+    $filter_domain = array();
+    for ($i=0;$i<$result->num_rows;$i++) {
+        $filter = $result->fetch_row();
+        $filter_domain[] = $filter[0];
+    }
+    if ($_SESSION['user_type'] === 'D' && count($parts) === 1 && $_SESSION['domain'] !== '') {
+        return getHtmlMessage(__('error'.$method.'nodomainforbidden12'), 'error');
+    } elseif ($_SESSION['user_type'] === 'D' && count($parts) === 2 && ($parts[1] !== $_SESSION['domain'] && in_array($parts[1], $filter_domain, true) === false)) {
+        return getHtmlMessage(sprintf(__('error'.$method.'domainforbidden12'), $parts[1]), 'error');
+    }
+}
+
+/**
+ * @param string $username
+ * @param string $userType
+ * @param string $oldUserType
+ */
+function testPermissions($username, $userType, $oldUserType)
+{
+    if (($_SESSION['user_type'] !== 'A' && $oldUserType === 'A')|| $_SESSION['user_type'] === 'D' && $_SESSION['myusername'] !== $username && $userType !== 'U' && (!defined('ENABLE_SUPER_DOMAIN_ADMINS') || ENABLE_SUPER_DOMAIN_ADMINS === false)) {
+        return getHtmlMessage(__('erroradminforbidden12'), 'error');
+    } elseif ($_SESSION['user_type'] === 'D' && $userType === 'A') {
+        return getHtmlMessage(__('errortypesetforbidden12'), 'error');
+    }
+}
+
+/**
+ * @param string $username
+ * @param string $usertype
+ * @param string $oldUsername
+ */
+function testValidUser($username, $usertype, $oldUsername)
+{
+    if ($usertype !== 'A' && validateInput($username, 'email') === false && (!defined('ALLOW_NO_USER_DOMAIN') || ALLOW_NO_USER_DOMAIN === false)) {
+        return getHtmlMessage(__('forallusers12'), 'error');
+    } elseif (!isset($_POST['password'], $_POST['password1'])) {
+        return getHtmlMessage(__('dievalidate99'), 'error');
+    } elseif ($_POST['password'] === '') {
+        return getHtmlMessage(__('errorpwdreq12'), 'error');
+    } elseif ($_POST['password'] !== $_POST['password1']) {
+        return getHtmlMessage(__('errorpass12'), 'error');
+    } elseif ($username === '') {
+        return getHtmlMessage(__('erroruserreq12'), 'error');
+    } elseif ($oldUsername !== $username && checkForExistingUser($username)) {
+        return getHtmlMessage(sprintf(__('userexists12'), sanitizeInput($username)), 'error');
+    }
+}
+
+function testtoken()
+{
+    if (!isset($_POST['token']) && !isset($_GET['token'])) {
+        return getHtmlMessage(__('dievalidate99'), 'error');
+    }
+    if ((isset($_POST['token']) && (false === checkToken($_POST['token'])))
+          || (isset($_GET['token']) && (false === checkToken($_GET['token'])))) {
+        return getHtmlMessage(__('dietoken99'), 'error');
+    }
+}
+
+function getUserById($additionalFields = false)
+{
+    if (isset($_POST['id'])) {
+        $uid = $_POST['id'];
+    } elseif (isset($_GET['id'])) {
+        $uid = $_GET['id'];
```

Comment:
`$uid = (int)$_GET['id'];` will make a first quick filter on user input (is is always integer)